Mathlib/Tactic/Says.lean

```diff
@@ -0,0 +1,115 @@
+    Except String (TSyntax ``tacticSeq) :=
+  let p := andthenFn whitespace Tactic.tacticSeq.fn
+  let ictx := mkInputContext input fileName
+  let s := p.run ictx { env, options := {} } (getTokenTable env) (mkParserState input)
+  if s.hasError then
+    Except.error (s.toErrorMsg ictx)
+  else if input.atEnd s.pos then
+    Except.ok ⟨s.stxStack.back⟩
+  else
+    Except.error ((s.mkError "end of input").toErrorMsg ictx)
+
+/--
+Run `evalTactic`, capturing any new messages.
+The optional `only` argument allows selecting which messages should be captured,
+or left in the message log.
+-/
+def evalTacticCapturingMessages (tac : TSyntax `tactic) (only : Message → Bool := fun _ => true) :
+    TacticM (List Message) := do
+  let mut msgs ← modifyGetThe Core.State fun st => (st.messages, { st with messages := {} })
+  try
+    evalTactic tac
+    let (capture, leave) := (← getThe Core.State).messages.msgs.toList.partition only
+    msgs := ⟨leave.foldl (fun m => m.push) msgs.msgs⟩
+    return capture
+  catch e =>
+    msgs := msgs ++ (← getThe Core.State).messages
+    throw e
+  finally
+    modifyThe Core.State fun st => { st with messages := msgs }
+
+/--
+Run `evalTactic`, capturing any new info messages.
+-/
+def evalTacticCapturingInfo (tac : TSyntax `tactic) : TacticM (List Message) :=
+  evalTacticCapturingMessages tac fun m => match m.severity with | .information => true | _ => false
+
+/--
+Run `evalTactic`, capturing a "Try this:" message and converting it back to syntax.
+-/
+def evalTacticCapturingTryThis (tac : TSyntax `tactic) : TacticM (TSyntax ``tacticSeq) := do
+  let msg ← match ← evalTacticCapturingInfo tac with
+  | [] => throwError m!"Tactic `{tac}` did not produce any messages."
+  | [msg] => msg.toString
+  | _ => throwError m!"Tactic `{tac}` produced multiple messages."
+  let tryThis ← match msg.dropPrefix? "Try this: " with
+  | none => throwError m!"Tactic output did not begin with 'Try this:': {msg}"
+  | some S => pure S.toString
+  match parseAsTacticSeq (← getEnv) tryThis with
+  | .ok stx => return stx
+  | .error err => throwError m!"Failed to parse tactic output: {tryThis}\n{err}"
+
+syntax (name := says) tactic " says" (colGt tacticSeq)? : tactic
+
+elab_rules : tactic
+  | `(tactic| $tac:tactic says%$tk $[$result:tacticSeq]?) => do
+  let verify := says.verify.get (← getOptions) ||
+    !says.no_verify_in_CI.get (← getOptions) && (← IO.getEnv "CI").isSome
+  match result, verify with
+  | some _, true
+  | none, _ =>
+    let stx ← evalTacticCapturingTryThis tac
+    match result with
+    | some r =>
+        let stx' := (← Lean.PrettyPrinter.ppTactic ⟨stx⟩).pretty
+        let r' := (← Lean.PrettyPrinter.ppTactic ⟨r⟩).pretty
+        if stx' != r' then
+          throwError m!"Tactic `{tac}` produced `{stx'}`,\nbut was expecting it to produce `{r'}`!"
+    | none =>
+    addSuggestion tk (← `(tactic| $tac says $stx)) (origSpan? := (← `(tactic| $tac says)))
```

Comment:
I suspect this should be something like
```suggestion
    addSuggestion tk (← `(tactic| $tac says $stx)) (origSpan? := (← getRef))
```
but I'm not sure there's any observable difference since quasiquotes use `getRef`.

Let's merge and see what `says` is like in practice.

bors r+